source/geod24/Serialization.d

```diff
@@ -0,0 +1,259 @@
+    Other utilities in this module are used to facilitate development
+    and diagnostics.
+
+    Author:         Mathias 'Geod24' Lang
+    License:        MIT (See LICENSE.txt)
+    Copyright:      Copyright (c) 2020 Mathias Lang. All rights reserved.
+
+*******************************************************************************/
+
+module geod24.Serialization;
+
+import std.traits;
+
+/*******************************************************************************
+
+    Serialize arbitrary data to / from JSON
+
+    This is the default serializer used by LocalRest, when none is provided.
+    It is a template so that LocalRest does not import Vibe.d if not used.
+
+*******************************************************************************/
+
+public template VibeJSONSerializer ()
+{
+    import vibe.data.json;
+
+    public string serialize (T) (auto ref T value) @safe
+    {
+        return serializeToJsonString(value);
+    }
+
+    public QT deserialize (QT) (string data) @safe
+    {
+        return deserializeJson!(QT)(data);
+    }
+}
+
+///
+unittest
+{
+    alias S = VibeJSONSerializer!();
+    static assert(!serializerInvalidReason!(S).length, serializerInvalidReason!(S));
+    static assert(is(SerializedT!(S) == string));
+}
+
+/// Utility function to check if a serializer conforms to the requirements
+/// Returns: An error message, or `null` if the serializer is conformant
+public string serializerInvalidReason (alias S) ()
+{
+    // Give codegen a break
+    if (!__ctfe) return null;
+
+    static if (!is(typeof(() { alias X = S.serialize; })))
+        return "`" ~ S.stringof ~ "` is missing a `serialize` method";
+    else static if (!is(typeof(() { alias X = S.deserialize; })))
+        return "`" ~ S.stringof ~ "` is missing a `deserialize` method";
+
+    /// We need a type to deserialize to test this function
+    /// While we should only test with return types / parameter types,
+    /// pretty much anything that can't (de)serialize an `int` is broken.
+
+    else static if (!is(typeof(() { S.serialize(0); })))
+        return "`" ~ S.stringof ~ ".serialize` is not callable using argument `int`";
+    else static if (!isCallable!(S.deserialize!int))
+        return "`" ~ S.stringof ~ ".deserialize` is not callable";
+
+    // If the template has an error we want to be informative
+    else static if (!is(typeof(S.serialize(0))))
+        return "`" ~ S.stringof ~ ".serialize(0)` does not return any value, does it compile?";
+
+    // All accepted types convert to `immutable(void)[]`
+    else static if (!is(typeof(S.serialize(0)) : immutable(void)[]))
+        return "`" ~ S.stringof ~ ".serialize` return value should be "
+        ~ "`string`, `immutable(ubyte)[]`, or `immutable(void)[]`, not: `"
+        ~ typeof(S.serialize(0)).stringof ~"`";
+
+    else
+    {
+        // Actual return type used
+        alias RT = SerializedT!S;
+
+        static if (!is(typeof(S.deserialize!int(RT.init))))
```

Comment:
Same question about lvalues.

Reply:
I don't see a reason to take it by `ref` ? It's a const array.

Reply:
Never mind that question, it's an array.